### PR TITLE
{Keyvault} az keyvault security-domain upload, adding restore-blob to prevent exposing keys in online environment

### PR DESCRIPTION
**Pull request by rebremer**
**Related command**
az keyvault security-domain upload

**Description**
az keyvault security-domain upload combines the following two things: 1) decrypting/encrypting of security domain using high value keys and 2) uploading the encrypted security domain to restore the HSM. This implies that high value keys are exposed in an online domain. This proposed change splits the decrypting/encrypting of security domain in a separate az keyvault security-domain restore-blob command and then only use az keyvault security-domain upload encrypted blob using --restore-blob flag

**Testing Guide**
az keyvault security-domain init-recovery --hsm-name $hsm_name --sd-exchange-key ContosoMHSM2-SDE.cer
az keyvault security-domain restore-blob --sd-exchange-key ContosoMHSM2-SDE.cer --sd-file ContosoMHSM-SD.json --sd-wrapping-keys cert_0.key cert_1.key cert_2.key --sd-file-restore-blob restore_blob.json 
az keyvault security-domain upload --hsm-name $hsm_name --sd-file restore_blob.json --restore-blob

**History Notes**

[Keyvault] `az keyvault security-domain restore-blob`: Support restoring blob offline
[Keyvault] `az keyvault security-domain upload`: Add `--restore-blob` to prevent exposing keys in online environment

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
keyvault

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 1]
```diff
@@ -774,13 +774,26 @@
         az keyvault security-domain init-recovery --hsm-name MyHSM --sd-exchange-key "{PATH_TO_RESTORE}"
 """
 
+helps['keyvault security-domain restore-blob'] = """
+type: command
+short-summary: Enables to decrypt and encrypt security domain file as blob. Can be run in offline environment, before file is uploaded to HSM using security-domain upload.
+examples:
+  - name: Security domain restore blob).
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:781]:
```suggestion
short-summary: Enable to decrypt and encrypt security domain file as blob. Can be run in offline environment, before file is uploaded to HSM using security-domain upload.
examples:
  - name: Security domain restore blob.
```

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 2]
```diff
@@ -2506,10 +2511,13 @@ def security_domain_upload(cmd, client, hsm_name, sd_file, sd_exchange_key, sd_w
         shared_keys=shared_keys,
         required=required
     )
+    return restore_blob_value
+
+def _security_domain_upload_blob(cmd, client, hsm_name, restore_blob_value, identifier=None, vault_base_url=None, no_wait=False):
     SecurityDomainObject = cmd.get_models('SecurityDomainObject', resource_type=ResourceType.DATA_PRIVATE_KEYVAULT)
     security_domain = SecurityDomainObject(value=restore_blob_value)
+    print(security_domain)
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:2519]: Is this `print` for debug? Do you want to keep it in released CLI?

**Review by evelyn-ys — COMMENTED** [review]
Can you add tests for the new command and new parameter?

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py (hunk 2 above)
Comment (rebremer) [reply src/azure-cli/azure/cli/command_modules/keyvault/custom.py]: Print is for debug, bugfix,  I removed it

File: src/azure-cli/azure/cli/command_modules/keyvault/_help.py [hunk 3]
```diff
@@ -774,13 +774,26 @@
...
 helps['keyvault security-domain upload'] = """
 type: command
 short-summary: Start to restore the HSM.
 examples:
   - name: Security domain upload (M=2).
     text: |
         az keyvault security-domain upload --hsm-name MyHSM --sd-file "{SD_TRANSFER_FILE}" --sd-exchange-key "{PEM_FILE_NAME}" --sd-wrapping-keys "{PEM_PRIVATE_KEY1_FILE_NAME}" "{PEM_PRIVATE_KEY2_FILE_NAME}"
+  - name: Security domain upload, in which sd_file is already restored using keyvault security-domain restore-blob command
+    text: |
+        az keyvault security-domain upload --hsm-name MyHSM --sd-file "{SD_TRANSFER_FILE}" --restore-blob"
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/keyvault/_help.py:796]:
```suggestion
        az keyvault security-domain upload --hsm-name MyHSM --sd-file "{SD_TRANSFER_FILE}" --restore-blob
```

File: src/azure-cli/azure/cli/command_modules/keyvault/_params.py [hunk 4]
```diff
@@ -583,13 +584,25 @@ class CLISecurityDomainOperation(str, Enum):
     with self.argument_context('keyvault security-domain upload') as c:
         c.argument('sd_file', help='This file contains security domain encrypted using SD Exchange file downloaded '
                                    'in security-domain init-recovery command.')
+        c.argument('restore_blob', help='Indicator if blob is already restored.')
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/keyvault/_params.py:587]:
```suggestion
        c.argument('restore_blob', help='Indicator if blob is already restored.')
```

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 5]
```diff
@@ -2468,8 +2468,13 @@ def _security_domain_gen_blob(sd_exchange_key, share_arrays, enc_data, required)
...
 
 
-def security_domain_upload(cmd, client, hsm_name, sd_file, sd_exchange_key, sd_wrapping_keys, passwords=None,
-                           identifier=None, vault_base_url=None, no_wait=False):  # pylint: disable=unused-argument
+def _security_domain_restore_blob(sd_file, sd_exchange_key, sd_wrapping_keys, passwords=None):
+    
+    if sd_exchange_key == None:
+        raise RequiredArgumentMissingError('Please specify --sd_exchange_key')
+    if sd_wrapping_keys == None:
+        raise RequiredArgumentMissingError('Please specify --sd_wrapping_keys')
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:2476]:
```suggestion
    if not sd_exchange_key:
        raise RequiredArgumentMissingError('Please specify --sd-exchange-key')
    if not sd_wrapping_keys:
        raise RequiredArgumentMissingError('Please specify --sd-wrapping-keys')
```

File: src/azure-cli/azure/cli/command_modules/keyvault/custom.py [hunk 6]
```diff
@@ -2520,6 +2527,27 @@ def security_domain_upload(cmd, client, hsm_name, sd_file, sd_exchange_key, sd_w
         return new_retval
     return retval
 
+def security_domain_upload(cmd, client, hsm_name, sd_file, restore_blob=False, sd_exchange_key=None, sd_wrapping_keys=None, passwords=None,
+                           identifier=None, vault_base_url=None, no_wait=False):  # pylint: disable=unused-argument
+    if restore_blob == False:
```
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/keyvault/custom.py:2532]:
```suggestion
    if not restore_blob:
```

**Pull request merged**